### ISSUE TITLE
Rust doesn't warn about never read fields in unreachable pub structures

### ISSUE THREAD

**Issue by xfix** [rust-lang/rust#97965]
Given the following code: 

```rust
mod m {
    pub struct Foo {
        pub a: i32,
        pub b: i32,
    }
}
fn main() {
    let _x = m::Foo { a: 3, b: 4 };
}
```

The current output is:

```
```

Ideally the output should look like:

```
warning: field is never read: `a`
 --> src/main.rs:3:9
  |
3 |         pub a: i32,
  |         ^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: field is never read: `b`
 --> src/main.rs:4:9
  |
4 |         pub b: i32,
  |         ^^^^^^^^^^
```

When using `pub(crate)` the warnings appear, which suggests to me that the same should be the case when having unreachable `pub`.

```rust
mod m {
    pub(crate) struct Foo {
        pub a: i32,
        pub b: i32,
    }
}
fn main() {
    let _x = m::Foo { a: 3, b: 4 };
}
```
